eras/babbage/impl/src/Cardano/Ledger/Babbage/Collateral.hs

```diff
@@ -0,0 +1,92 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+
+-- | Figure 2: Functions related to fees and collateral
+--   Babbage Specification
+module Cardano.Ledger.Babbage.Collateral where
+
+import Cardano.Ledger.Address (Addr (..))
+import Cardano.Ledger.Alonzo.TxWitness (TxWitness)
+import Cardano.Ledger.Babbage.Scripts (txscripts)
+import Cardano.Ledger.Babbage.TxBody
+  ( TxBody (..),
+    TxOut (..),
+    collateralInputs',
+    collateralReturn',
+    outputs',
+    txfee',
+  )
+import Cardano.Ledger.Coin (Coin (..))
+import qualified Cardano.Ledger.Core as Core
+import Cardano.Ledger.Era (Era (Crypto), ValidateScript (..))
+import Cardano.Ledger.Shelley.UTxO (UTxO (..), balance, getScriptHash)
+import Cardano.Ledger.TxIn (TxIn (..), txid)
+import Cardano.Ledger.Val ((<->))
+import Data.Compact.SplitMap ((◁))
+import qualified Data.Compact.SplitMap as SplitMap
+import Data.Maybe.Strict (StrictMaybe (..))
+import GHC.Records (HasField (..))
+import Numeric.Natural (Natural)
+
+-- ============================================================
+
+isTwoPhaseScriptAddress ::
```

Comment:
can we not reuse the version from Alonzo? https://github.com/input-output-hk/cardano-ledger/blob/2048cd4679ef3155d88a445e0f818abad09a0827/eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs#L268

Reply:
nevermind! it's slightly different in babbage